src/Data/Vec/Functional.agda

```diff
@@ -0,0 +1,82 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- Vectors defined via index notation
+------------------------------------------------------------------------
+
+{-# OPTIONS --without-K --safe #-}
+
+module Data.Vec.Functional where
+
+open import Data.Fin using (Fin; zero; suc)
+open import Data.Nat using (ℕ; zero; suc; _+_)
+open import Data.Product using (Σ; ∃; _×_; _,_; proj₁; proj₂)
+open import Data.Vec as V using (Vec)
+open import Function
+open import Level using (Level)
+
+infixr 5 _∷_ _++_
+infixl 4 _⊛_
+
+private
+  variable
+    a b c : Level
+    A : Set a
+    B : Set b
+    C : Set c
+
+Vector : Set a → ℕ → Set a
+Vector A n = Fin n → A
+
+map : (A → B) → ∀ {n} → (Vector A n → Vector B n)
+map f xs = f ∘ xs
+
+rearrange : ∀ {m n} → (Fin m → Fin n) → (Vector A n → Vector A m)
+rearrange r xs = xs ∘ r
+
+[] : Vector A zero
+[] ()
+
+_∷_ : ∀ {n} → A → Vector A n → Vector A (suc n)
+(x ∷ xs) zero = x
+(x ∷ xs) (suc i) = xs i
+
+head : ∀ {n} → Vector A (suc n) → A
+head xs = xs zero
+
+tail : ∀ {n} → Vector A (suc n) → Vector A n
+tail xs = xs ∘ suc
+
+uncons : ∀ {n} → Vector A (suc n) → A × Vector A n
+uncons xs = head xs , tail xs
+
+_++_ : ∀ {m n} → Vector A m → Vector A n → Vector A (m + n)
```

Comment:
It feels like append should be factored through the isomorphism
`Fin (m + n) ~= Fin m + Fin n`.

Reply:
As far as I can see, we don't have this isomorphism in stdlib. We *really* should have it, so I'll take this as an excuse to implement it.

It changes the computational behaviour, but possibly for the better.

Reply:
Yes, it should. And I have those proofs in https://github.com/JacquesCarette/pi-dual/blob/master/Univalence/FinEquivPlusTimes.agda that I've been meaning to port & submit to stdlib. This might be the push I needed.

Reply:
@laMudri go ahead and 'borrow' any parts of my code if it helps. I wasn't necessarily going to do it today, so it sounds like you'll get to it sooner.